modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C

```diff
@@ -169,350 +226,319 @@ ADLAROMANCEStressUpdateBase::initQpStatefulProperties()
-                       ")");
-        else if (_window_failure == WindowFailure::ERROR)
-          mooseError("In ",
-                     _name,
-                     ": input parameter number input=",
-                     j,
-                     " output=",
-                     i,
-                     " with value (",
-                     MetaPhysicL::raw_value(input[j]),
-                     ") is out of range (",
-                     _input_limits[j][0],
-                     " - ",
-                     _input_limits[j][1],
-                     "), window (",
-                     _window,
-                     ")");
+        smoother_step = MathUtils::smootherStep(input, 0.0, limits[0], derivative);
+        input = limits[0];
       }
+      else
+        input = limits[1];
+    }
+    else
+    {
+      std::stringstream msg;
+      msg << "In " << _name << ": input parameter with value (" << MetaPhysicL::raw_value(input)
+          << ") is out of range (" << limits[0] << " - " << limits[1] << ")";
+
+      if (behavior == WindowFailure::WARN)
+        mooseWarning(msg.str());
+      else if (behavior == WindowFailure::ERROR)
+        mooseError(msg.str());
     }
   }
+
+  return smoother_step;
 }
 
-void
-ADLAROMANCEStressUpdateBase::convertInput(const std::vector<ADReal> & input,
-                                          std::vector<std::vector<ADReal>> & converted,
-                                          std::vector<std::vector<ADReal>> & dconverted)
+ADReal
+ADLAROMANCEStressUpdateBase::convertInput(const ADReal & input,
+                                          const ROMInputTransform transform,
+                                          const Real transform_coef,
+                                          const std::vector<Real> & transformed_limits,
+                                          const bool jacobian)
 {
-  for (unsigned int i = 0; i < _num_outputs; ++i)
+  ADReal x(1.0);
+  if (transform == ROMInputTransform::LINEAR)
   {
-    for (unsigned int j = 0; j < _num_inputs; ++j)
-    {
-      ADReal x = input[j];
-      ADReal dx = 1.0;
-      if (_transform[i][j] == ROMInputTransform::EXP)
-      {
-        x = std::exp(x / _transform_coefs[i][j]);
-        dx = x / _transform_coefs[i][j];
-      }
-      else if (_transform[i][j] == ROMInputTransform::LOG)
-      {
-        dx = 1.0 / (x + _transform_coefs[i][j]);
-        x = std::log(x + _transform_coefs[i][j]);
-      }
-
-      converted[i][j] = 2.0 * (x - _transformed_limits[i][j][0]) /
-                            (_transformed_limits[i][j][1] - _transformed_limits[i][j][0]) -
-                        1.0;
-      if (j == _stress_index)
-        dconverted[i][j] = dx * 2.0 / (_transformed_limits[i][j][1] - _transformed_limits[i][j][0]);
-      else
-        dconverted[i][j] = 0.0;
-    }
+    if (!jacobian)
+      x = input;
+  }
+  else if (transform == ROMInputTransform::EXP)
+  {
+    if (!jacobian)
+      x = std::exp(input / transform_coef);
```

Comment:
Probably want to check for divide by zero here. Also, I'm not really well versed in AD -- does transform_coef need to be an ADReal, or is it not necessary because it wouldn't change, and therefore doesn't affect the Jacobian?

Reply:
Yeah, exactly about ADReal/Real. I'll add a check for transform_coef in the initialize stage, good point.